psbt: Only include PSBT_OUT_TAP_TREE when the output has a script path

achow101 opened the pull request:
PSBT_OUT_TAP_TREE should not be included for outputs that do not have such a tree. This should be disallowed during parsing, as well as prior to serialization when the field is populated during updating.

Also added some test cases.

Alternative to #25856

JeremyRubin committed:
[BugFix]: Do not allow deserializing PSBT with empty PSBT_OUT_TAP_TREE

achow101 committed:
psbt: Only include m_tap_tree if it has scripts

achow101 committed:
tests: Test that PSBT_OUT_TAP_TREE is included correctly

JeremyRubin commented:
Either approach should work.

Not as big a fan of this approach because we may, in the future, wish to have access to the TaprootBuilder even HasScripts is false. E.g., were the TaprootBuilder to be extended to contain information about MuSig derivations, in which case the changes would be more straightforward IMO.

Of course, it's silly to predict how future changes might be made, but I just personally find it conceptually cleaner to hang onto whatever data the sigdata passed us, and to use it correctly on output.

achow101 commented:
> Not as big a fan of this approach because we may, in the future, wish to have access to the TaprootBuilder even HasScripts is false. E.g., were the TaprootBuilder to be extended to contain information about MuSig derivations, in which case the changes would be more straightforward IMO.

I highly doubt that would be the case. And if it were, it's not particularly difficult to deal with it then. Furthermore, the general structure of the PSBT implementation is that each field has its own explicit member inside its struct, so any future things would get their own fields, which will then be un/packed from TaprootBuilder as needed.

Perhaps the more correct way to implement `m_tap_tree` is to make it the vector of tuples rather than a TaprootBuilder.

achow101 committed:
psbt: Change m_tap_tree to store just the tuples

Instead of having an entire TaprootBuilder which may or may not be
complete, and could potentially have future changes that interact oddly
with taproot tree tuples, have m_tap_tree be just the tuples.

When needed in other a TaprootBuilder for actual use, the tuples will be
added to a a TaprootBuilder that, in the future, can take in whatever
other data is needed as well.

achow101 commented:
> Perhaps the more correct way to implement `m_tap_tree` is to make it the vector of tuples rather than a TaprootBuilder.

Added a commit that does this.

JeremyRubin commented on src/psbt.cpp:
```diff
@@ -219,7 +219,16 @@ void PSBTOutput::FillSignatureData(SignatureData& sigdata) const
+        }
+        assert(builder.IsComplete());
+        builder.Finalize(m_tap_internal_key);
+        TaprootSpendData spenddata = builder.GetSpendData();
```
maybe this should just be a function in TaprootBuilder?

JeremyRubin commented:
approach ACK, this seems OK.

darosior commented on src/psbt.h:
```diff
@@ -713,7 +713,7 @@ struct PSBTOutput
     std::map<CPubKey, KeyOriginInfo> hd_keypaths;
     XOnlyPubKey m_tap_internal_key;
-    std::optional<TaprootBuilder> m_tap_tree;
+    std::optional<std::vector<std::tuple<uint8_t, uint8_t, CScript>>> m_tap_tree;
```
nit: What's the reason for keeping it optional now it's a vector?

achow101 replied:
Dropped the optional

darosior commented on test/functional/rpc_psbt.py:
```diff
@@ -834,6 +847,9 @@ def test_psbt_input_keys(psbt_input, keys):
         assert_raises_rpc_error(-8, "PSBTs not compatible (different transactions)", self.nodes[0].combinepsbt, [psbt1, psbt2])
         assert_equal(self.nodes[0].combinepsbt([psbt1, psbt1]), psbt1)
 
+        if self.options.descriptors:
+            self.log.info("Test that PSBT_OUT_TAPROOT_TREE is only present for Taproot outputs that have script paths")
+
```
Looks like you actually didn't commit this test?

achow101 replied:
Removed. I changed how I wanted to do the test and forgot to drop the log line.

darosior reviewed (commented):
Code looks good to me. Looks like part of the functional test is missing.

aureleoules commented on src/psbt.cpp:
```diff
@@ -219,7 +219,16 @@ void PSBTOutput::FillSignatureData(SignatureData& sigdata) const
     if (m_tap_tree.has_value() && m_tap_internal_key.IsFullyValid()) {
-        TaprootSpendData spenddata = m_tap_tree->GetSpendData();
+        TaprootBuilder builder;
+        for (const auto& tuple : m_tap_tree.value()) {
+            uint8_t depth = std::get<0>(tuple);
+            uint8_t leaf_ver = std::get<1>(tuple);
+            CScript script = std::get<2>(tuple);
```
```suggestion
        for (const auto& [depth, leaf_ver, script] : m_tap_tree.value()) {
```

achow101 replied:
Done

aureleoules commented on src/psbt.h:
```diff
@@ -758,8 +758,7 @@ struct PSBTOutput
             CVectorWriter s_value(s.GetType(), s.GetVersion(), value, 0);
-            const auto& tuples = m_tap_tree->GetTreeTuples();
-            for (const auto& tuple : tuples) {
+            for (const auto& tuple : m_tap_tree.value()) {
                 uint8_t depth = std::get<0>(tuple);
                 uint8_t leaf_ver = std::get<1>(tuple);
                 CScript script = std::get<2>(tuple);
```
```suggestion
            for (const auto& [depth, leaf_ver, script] : m_tap_tree.value()) {
```

achow101 replied:
Done

aureleoules commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1243,8 +1243,7 @@ static RPCHelpMan decodepsbt()
             UniValue tree(UniValue::VARR);
-            const auto& tuples = output.m_tap_tree->GetTreeTuples();
-            for (const auto& tuple : tuples) {
+            for (const auto& tuple : output.m_tap_tree.value()) {
                 uint8_t depth = std::get<0>(tuple);
                 uint8_t leaf_ver = std::get<1>(tuple);
                 CScript script = std::get<2>(tuple);
```
```suggestion
            for (const auto& [depth, leaf_ver, script] : output.m_tap_tree.value()) {
```

achow101 replied:
Done

achow101 commented on src/script/standard.h:
```diff
@@ -315,6 +315,8 @@ class TaprootBuilder
     /** Returns a vector of tuples representing the depth, leaf version, and script */
     std::vector<std::tuple<uint8_t, uint8_t, CScript>> GetTreeTuples() const;
+    /** Returns true if there are any tapscripts */
+    bool HasScripts() const { return m_branch.size() > 0; }
```
Done

S3RK commented:
Concept ACK

darosior commented:
ACK e04f5b93fb

JeremyRubin commented on src/psbt.cpp:
```diff
@@ -265,7 +271,7 @@ void PSBTOutput::Merge(const PSBTOutput& output)
     if (witness_script.empty() && !output.witness_script.empty()) witness_script = output.witness_script;
     if (m_tap_internal_key.IsNull() && !output.m_tap_internal_key.IsNull()) m_tap_internal_key = output.m_tap_internal_key;
-    if (m_tap_tree.has_value() && !output.m_tap_tree.has_value()) m_tap_tree = output.m_tap_tree;
+    if (!m_tap_tree.empty() && output.m_tap_tree.empty()) m_tap_tree = output.m_tap_tree;
```
isn't this the opposite of what we want?

why is the logic switched to "clear this m_tap_tree if merging output has none set" from "if unset, set m_tap_tree if output has one set".

darosior replied:
The logic is still "clear this m_tap_tree if merging output has none set". It's just that "has value" is the opposite of "is empty" so it had to be inversed.

JeremyRubin replied:
hmmm, even if not inverted, is the original behavior the intended behavior and why?

Other lines seem to be opposite, e.g. 273 has `set m_tap_key_internal from non-null output.m_tap_key_internal` semantics, not `unset m_tap_key_internal from null output.m_tap_key_internal`

```
if (m_tap_internal_key.IsNull() && !output.m_tap_internal_key.IsNull()) m_tap_internal_key = output.m_tap_internal_key;
```

JeremyRubin replied:
There should be a test covering this stuff as well?

achow101 replied:
Added a test.

darosior commented on test/functional/rpc_psbt.py:
```diff
@@ -779,9 +780,18 @@ def test_psbt_input_keys(psbt_input, keys):
+            parsed_psbt = PSBT.from_base64(psbt)
+            assert_greater_than(len(parsed_psbt.o[vout].map[PSBT_OUT_TAP_TREE]), 0)
+            assert "taproot_tree" in self.nodes[0].decodepsbt(psbt)["outputs"][vout]
+            parsed_psbt.make_blank()
+            comb_psbt = self.nodes[0].combinepsbt([psbt, parsed_psbt.to_base64()])
+            assert_equal(comb_psbt, psbt)
```
This does not quite test the bug Jeremy found. You can reintroduce it and the test won't fail
```diff
diff --git a/src/psbt.cpp b/src/psbt.cpp
index cbf2f88788..1567998039 100644
--- a/src/psbt.cpp
+++ b/src/psbt.cpp
@@ -271,7 +271,7 @@ void PSBTOutput::Merge(const PSBTOutput& output)
     if (redeem_script.empty() && !output.redeem_script.empty()) redeem_script = output.redeem_script;
     if (witness_script.empty() && !output.witness_script.empty()) witness_script = output.witness_script;
     if (m_tap_internal_key.IsNull() && !output.m_tap_internal_key.IsNull()) m_tap_internal_key = output.m_tap_internal_key;
-    if (m_tap_tree.empty() && !output.m_tap_tree.empty()) m_tap_tree = output.m_tap_tree;
+    if (!m_tap_tree.empty() && output.m_tap_tree.empty()) m_tap_tree = output.m_tap_tree;
 }
 bool PSBTInputSigned(const PSBTInput& input)
 {
```

achow101 replied:
Should work now with the `self.g` fix.

aureleoules replied:
cherry-picking 22c051ca70bae73e0430b05fb9d879591df27699 on master still doesn't make the test fail.

achow101 commented on src/psbt.cpp:
```diff
@@ -265,7 +271,7 @@ void PSBTOutput::Merge(const PSBTOutput& output)
     if (witness_script.empty() && !output.witness_script.empty()) witness_script = output.witness_script;
     if (m_tap_internal_key.IsNull() && !output.m_tap_internal_key.IsNull()) m_tap_internal_key = output.m_tap_internal_key;
-    if (m_tap_tree.has_value() && !output.m_tap_tree.has_value()) m_tap_tree = output.m_tap_tree;
+    if (m_tap_tree.empty() && !output.m_tap_tree.empty()) m_tap_tree = output.m_tap_tree;
```
Not really? It's part of the change of `m_tap_tree`'s type.

JeremyRubin replied:
i think what @darosior is suggesting is that this PR should have a "parent" PR which first fixes this bug?

the original line

```c++
    if (m_tap_tree.has_value() && !output.m_tap_tree.has_value()) m_tap_tree = output.m_tap_tree;
```

reads

replace the m_tap_tree if the incoming value is Empty and the current value is Something

which is clearly incorrect. now the PR is really two independent fixes, both of which might due with test coverage.

achow101 replied:
Ah, I see. I've split the fix and it's test into two separate commits.

achow101 commented on test/functional/test_framework/psbt.py:
```diff
@@ -123,6 +123,17 @@ def serialize(self):
+
+        old_g = self.g
+        self.g = PSBTMap()
+        self.g.map[0] = old_g.map[0]
```
Fixed.

achow101 committed:
psbt: Fix merging of m_tap_tree

Merging should be checking that the current PSBTOutput doesn't have a
taptree and the other one's is copied over. The original merging had
this inverted and would remove m_tap_tree if the other did not have it.

achow101 committed:
tests: Test that PSBT_OUT_TAP_TREE is combined correctly

instagibbs commented on src/psbt.cpp:
```diff
@@ -249,7 +249,7 @@ void PSBTOutput::FromSignatureData(const SignatureData& sigdata)
     if (!sigdata.tr_spenddata.internal_key.IsNull()) {
         m_tap_internal_key = sigdata.tr_spenddata.internal_key;
     }
-    if (sigdata.tr_builder.has_value()) {
```
let's add a test for this specific case

achow101 replied:
I think this is not really testable since we would not have a `tr_builder` for change that doesn't have scripts. This could only have been reached through a bad psbt, which have now been disallowed through serialization, and this change is more of a belt-and-suspenders.

instagibbs replied:
seems this would only be hit internally, resulting in a psbt with less data. not too troubling

darosior reviewed (approved):
ACK 9e386afb67bf8fa71b72f730da1695eeb11828cd

instagibbs commented:
ACK https://github.com/bitcoin/bitcoin/pull/25858/commits/9e386afb67bf8fa71b72f730da1695eeb11828cd

fanquake commented:
Added to #26133 for backport to 24.x.

JeremyRubin commented:
post ack, looks correct to me now
